### DISCUSSION TITLE
Feat request: support for Vue 2

### DISCUSSION THREAD

**Discussion by ziadghalleb** [tailwindlabs/headlessui#318] [Ideas]
It's unfortunate the headlessUI library is only available for Vue 3. I understand you want your framework to stay up to date but the adoption is all around low for Vue 3.

**Comment by adamwathan** [comment]
Yeah we definitely didn't expect Vue 3 adoption to be this poor this long after the release or we might have approached things differently. It's possible we do a Vue 2 version one day but it's a very big project and not something we can commit to yet. Every time this comes up I'm convinced that by the time we'd be able to pursue it, Vue 3 will have become the standard, but at this point I'm not sure that Vue 2 won't just be the de facto version forever. No promises but definitely paying attention 👍🏻

**Reply by tylerforesthauser** [reply]
Would you guys consider a compromise with the use of a library like vue-demi (https://github.com/vueuse/vue-demi)?

**Reply by adamwathan** [reply]
Still a monumental amount of work, the entire library is render functions and the render function API is totally different between Vue 2 and Vue 3, plus we use `teleport` which doesn't exist in Vue 2 (without a library, that has a different API), and there are limitations in the composition API plugin that would impact us as well. So either way I think it would feel mostly like writing two separate libraries.

**Reply by vesper8** [reply]
I plan to continue to use Vue2 for a long time. It would be really nice if I could make use of this new addition to TailwindUI.

I personally wouldn't mind a sort of translation layer so that you don't have to redo the whole thing using Vue2's render API.

It may be a little slower that way but it should mean a lot less effort to get it working asap. So it could go something like `if Vue2 then use https://github.com/LinusBorg/portal-vue else use teleport`. And then have a single place in headlessUI where you're making use of Vue3's render function and add a condition there `if Vue2 then use this render function else business as usual`

I'll admit I haven't looked at the code so don't actually know if a quick patch like this would work, it also depends if you've already created a kind of wrapper around Vue3's render function so you could add this workaround in a single centralized wrapper of sorts.

**Reply by divine** [reply]
@adamwathan 👋,

At first I thought why you guys are giving so much love for React but no love for Vue at all was so sad.

Anyways, according to this Vue RFC [comments](https://github.com/vuejs/rfcs/discussions/296#discussioncomment-562684):

> Composition API into Vue 2 core, then it will probably be around end of Q3 2021

But it will be another major release Vue 2.7.

Basically, it will make things much easier and I think that adding another set of components will be much better in a long run instead of creating a compatible library for Vue 2.6 which eventually will be replaced by Vue 2.7.

Thanks!

**Reply by cjlaborde** [reply]
Making this project for vue 3 instead of vue2 was a terrible decision. Considering vue2 is what the majority of vue user use and that nuxt.js only uses vue2 under the hood. Vue2 will still be more popular than vue3 for the next 2 years. Since vue 2.7 which has not been released yet will be the next version with LTS support for 18 months. That is a Long time and will probably make a lot of vue users like me ask for a refund since we mainly use nuxt.js or still maintain vue2 projects.

**Reply by tnorthcutt** [reply]
@cjlaborde what do you mean a "refund"? Headless UI is _free_.

**Reply by gustawdaniel-acaisoft** [reply]
I still use vue2 because of nuxt and I love vue2. In vue3 there are too many breaking changes. Please consider support for vue2. Thank you.

**Reply by jeremylowery** [reply]
@tnorthcutt He probably bought the license, which I did as well. At the time I had no idea this didn't support Vue2. Was a silly oversight by me. I keep checking back every month or so to see if it's supported.

**Reply by tremendus** [reply]
@adamwathan  - now that Vue 2.7+ is released - which backports (most of) the composition API to Vue 2 - is this perhaps feasible now? Has anyone from the Tailwind team had a chance to review and comment on this? Clearly there is demand ...

**Reply by adamwathan** [reply]
Don’t think it’s feasible because of all the breaking changes to the VNode APIs in Vue 3. No realistic chance we ever work on this I’m afraid sorry — even if a lot of people want it we simply don’t want to work on it, there’s other things to do that just sound a lot more fun and ultimately I just want to enjoy how I spend my time. It’s all MIT licensed open source though so anyone excited about this problem is of course free to use our work as a starting point and build a Vue 2 library themselves 👍🏻
